File: src/distortion_correction.cpp
#include "opencv_all.hpp"

int main(void)
{
    cv::Mat K = (cv::Mat_<double>(3, 3) << 432.7390364738057, 0, 476.0614994349778, 0, 431.2395555913084, 288.7602152621297, 0, 0, 1);
    cv::Mat dist_coeff = (cv::Mat_<double>(5, 1) << -0.2852754904152874, 0.1016466459919075, -0.0004420196146339175, 0.0001149909868437517, -0.01803978785585194);

    // Open an video
    cv::VideoCapture video;
    if (!video.open("data/chessboard.avi")) return -1;

    // Run distortion correction
    bool show_rectify = true;
    cv::Mat map1, map2;
    while (true)
    {
        // Grab an image from the video
        cv::Mat image;
        video >> image;
        if (image.empty()) break;

        // Rectify geometric distortion (c.f. 'cv::undistort()' can be applied for one-time remapping.)
        if (show_rectify)
        {
            if (map1.empty() || map2.empty())
                cv::initUndistortRectifyMap(K, dist_coeff, cv::Mat(), cv::Mat(), image.size(), CV_32FC1, map1, map2);
            cv::remap(image, image, map1, map2, cv::InterpolationFlags::INTER_LINEAR);
        }

        // Show the image
        cv::imshow("3DV Tutorial: Distortion Correction", image);
        int key = cv::waitKey(1);
        if (key == 27) break;                                   // 'ESC' key
        else if (key == 9) show_rectify = !show_rectify;        // 'Tab' key
        else if (key == 32)                                     // 'Space' key
        {
            key = cv::waitKey();
            if (key == 27) break;                               // 'ESC' key
            else if (key == 9) show_rectify = !show_rectify;    // 'Tab' key
        }
    }

    video.release();
    return 0;
}

File: calibration/distortion_correction.py
from re import S
import cv2
import numpy as np
import yaml
from yaml.loader import SafeLoader

def main():
    input = "data/chessboard.avi"
    cam_params_file = "cam_config.yaml"
    K = np.zeros((3,3))
    dist_coeff = np.zeros((4,1))
    with open(cam_params_file, 'r') as f:
        cam_params = yaml.load(f, Loader=SafeLoader)
        K = np.resize(cam_params["Intrinsic"], (3,3))
        dist_coeff = np.resize(cam_params["Distortion"], (4,1))

    # Open Video
    capture = cv2.VideoCapture(input)
    if capture.isOpened() == False: raise Exception("No video")

    show_rectify = True
    map1 = np.array([])
    map2 = np.array([])

    while True:
        ret, image = capture.read()
        if not ret: break
        
        info = "Original"
        if(show_rectify):
            if not (len(map1) & len(map2)):
                map1, map2 = cv2.initUndistortRectifyMap(K, dist_coeff, None, None, (image.shape[1], image.shape[0]), cv2.CV_32FC1)
            image = cv2.remap(image, map1, map2, interpolation=cv2.INTER_LINEAR)
            info = "Rectified"
        cv2.putText(image, info, (5,15), cv2.FONT_HERSHEY_PLAIN, 1, (0, 255,0))

        cv2.imshow("3DV Tutorial: Distortion Correction", image)

        key = cv2.waitKey(1)
        if key == 27: break
        elif key == 9: show_rectify = not show_rectify
        elif key == 32:
            key = cv2.waitKey()
            if key == 27: break
            elif key == 9: show_rectify = not show_rectify

    capture.release()

if __name__ == "__main__":
    main()